### PR TITLE
Expose signed distance sign type

### PR DESCRIPTION
**Pull request by daviesthomas**
This PR enables control of the method used to sign distances when calling `signed_distance(*)`

Adds additional parameter, `sign_type`, to `signed_distance(*)`, which allows the user to select one of 5 sign types outlined in igl/helpers.py. 

Valid usage of modified function:
``` python
# default to pseudonormal
S,I,C = signed_distance(P,V,F)
# return normals
S,I,C,N = signed_distance(P,V,F,return_normals=True)
# fast winding for non-manifold mesh
S,I,C = signed_distance(P,V,F,sign_type=SIGNED_DISTANCE_TYPE_FAST_WINDING_NUMBER)
```
Basic tests added to tests/test_basic.py to ensure:

- Each sign_type returns correctly formatted data
- Normals only returned when requested
- Assert raised if invalid sign type used when requesting normals
- Assert raised if invalid sign type given

(did not find PR guide so please let me know edits required to conform)

### CODE REVIEW

**Comment by teseoch** [comment]
looks good, can you add a check on `c_contiguous` and `dtype` of the output?

**Comment by teseoch** [comment]
can you check the `dtype` with the input, something like `self.v1.dtype`

**Comment by teseoch** [comment]
some of the assertions are failing :( you can check the ci log

**Comment by daviesthomas** [comment]
> some of the assertions are failing :( you can check the ci log

```
test_signed_distance (__main__.TestBasic) ... python: /home/travis/build/libigl/libigl-python-bindings/external/libigl/include/igl/signed_distance.cpp:73: void igl::signed_distance(const Eigen::MatrixBase<Derived>&, const Eigen::MatrixBase<U>&, const Eigen::MatrixBase<OtherDerived>&, igl::SignedDistanceType, typename DerivedV::Scalar, typename DerivedV::Scalar, Eigen::PlainObjectBase<DerivedD>&, Eigen::PlainObjectBase<DerivedsqrD>&, Eigen::PlainObjectBase<DerivedL>&, Eigen::PlainObjectBase<Derivedb>&) [with DerivedP = Eigen::Matrix<double, -1, -1>; DerivedV = Eigen::Matrix<double, -1, -1>; DerivedF = Eigen::Matrix<int, -1, -1>; DerivedS = Eigen::Matrix<double, -1, -1>; DerivedI = Eigen::Matrix<long int, -1, -1, 1, -1, -1>; DerivedC = Eigen::Matrix<double, -1, -1, 1, -1, -1>; DerivedN = Eigen::Matrix<double, -1, -1, 1, -1, -1>; typename DerivedV::Scalar = double]: Assertion `false && "Unknown SignedDistanceType"' failed.
```

This assertion error seems to be due to an outdated version of libigl relative to latest master. The assertions succeed on my machine (and I'd guess on all other builds). 

Do you have an idea why this might be happening or am I doing something wrong on my end.

**Comment by daviesthomas** [comment]
> > some of the assertions are failing :( you can check the ci log
> 
> ```
> test_signed_distance (__main__.TestBasic) ... python: /home/travis/build/libigl/libigl-python-bindings/external/libigl/include/igl/signed_distance.cpp:73: void igl::signed_distance(const Eigen::MatrixBase<Derived>&, const Eigen::MatrixBase<U>&, const Eigen::MatrixBase<OtherDerived>&, igl::SignedDistanceType, typename DerivedV::Scalar, typename DerivedV::Scalar, Eigen::PlainObjectBase<DerivedD>&, Eigen::PlainObjectBase<DerivedsqrD>&, Eigen::PlainObjectBase<DerivedL>&, Eigen::PlainObjectBase<Derivedb>&) [with DerivedP = Eigen::Matrix<double, -1, -1>; DerivedV = Eigen::Matrix<double, -1, -1>; DerivedF = Eigen::Matrix<int, -1, -1>; DerivedS = Eigen::Matrix<double, -1, -1>; DerivedI = Eigen::Matrix<long int, -1, -1, 1, -1, -1>; DerivedC = Eigen::Matrix<double, -1, -1, 1, -1, -1>; DerivedN = Eigen::Matrix<double, -1, -1, 1, -1, -1>; typename DerivedV::Scalar = double]: Assertion `false && "Unknown SignedDistanceType"' failed.
> ```
> 
> This assertion error seems to be due to an outdated version of libigl relative to latest master. The assertions succeed on my machine (and I'd guess on all other builds).
> 
> Do you have an idea why this might be happening or am I doing something wrong on my end.

see this PR which was merged a few weeks back: https://github.com/libigl/libigl/pull/1630

**Comment by teseoch** [comment]
can you update the libigl version?
https://github.com/libigl/libigl-python-bindings/blob/208aa3cfe80b06618c9b1a68267aab81181a36f8/cmake/PyiglDownloadExternal.cmake#L30

**Comment by daviesthomas** [comment]
> can you update the libigl version?
> https://github.com/libigl/libigl-python-bindings/blob/208aa3cfe80b06618c9b1a68267aab81181a36f8/cmake/PyiglDownloadExternal.cmake#L30

Will do. 

This unfortunatly requires some change to is_border_vertex(*) wrapper. The original function being called is now deprecated and removed (https://github.com/libigl/libigl/pull/1610) .

Only faces are required (not vertices) so I plan to allow vertices to be passed in still to avoid breaking existing users work but I can print a "DEPRECATED" notice. Or we can make the breaking change to refactor to:

"is_border_vertex(v,f)" --> "is_border_vertex(f)". 

Let me know how to proceed...

**Comment by teseoch** [comment]
I see your message in the email, not here, strange...

I think not breaking python is better.
I don't think printing is necessary, but if you belive it is better, go for the print :)

**Comment by daviesthomas** [comment]
Sorry I deleted it :)

I grabbed an older commit before the deprecated function removal PR in
libigl. This lets me keep my PR specific to the signed_distance code.

I don't mind submitting a separate PR updating the bindings to the latest
version with all the bindings updated to reflect changes (without breaking
api).

**Comment by daviesthomas** [comment]
I retract that. No such commit exists. I'll update all deprecated functions
as a part of this PR. I'll ensure no breaking changes to any function
calls.

File: cmake/PyiglDownloadExternal.cmake [hunk 1]
```diff
@@ -27,7 +27,7 @@ endfunction()
 function(pyigl_download_igl)
     pyigl_download_project(libigl
         GIT_REPOSITORY https://github.com/libigl/libigl.git
-        GIT_TAG        60ffe1b0dca692e6bee21d9be7e6fdefdb30d323
+        GIT_TAG        f06fb13014aadff7eb1d7423ac018620441a3657
```
Comment (daviesthomas) [inline cmake/PyiglDownloadExternal.cmake:30]: Corresponds to current latest commit on libigl: https://github.com/libigl/libigl/commit/f06fb13014aadff7eb1d7423ac018620441a3657

File: src/is_border_vertex.cpp [hunk 2]
```diff
@@ -40,7 +40,7 @@ npe_arg(f, dense_int, dense_long, dense_longlong)
 
 npe_begin_code()
   assert_valid_23d_tri_mesh(v, f);
-  const auto res = igl::is_border_vertex(v, f);
+  const auto res = igl::is_border_vertex(f);
```
Comment (daviesthomas) [inline src/is_border_vertex.cpp:43]: Change a result of this PR: https://github.com/libigl/libigl/pull/1610

Function taking V and F is deprecated. We don't change our (your 😄  ) API to ensure no breaking change. V is simply no longer used.

File: src/readMSH.cpp [hunk 3]
```diff
@@ -44,25 +37,26 @@ npe_arg(filename, std::string)
 npe_default_arg(dtypef, npe::dtype, "float")
 
 npe_begin_code()
+  Eigen::MatrixXd v;
+  Eigen::MatrixXi t;
+
+  //NOTE: readMSH only support vertices as matrix of doubles. We instead cast the output accordingly.
+  bool ret = igl::readMSH(filename, v, t);
```
Comment (daviesthomas) [inline src/readMSH.cpp:44]: readMSH refactored in PR: https://github.com/libigl/libigl/commit/8c95d1044ed509112108d168653ce7a1839c1508#diff-14358970970f78facdd59d4179315761ff7f7a3114d962c272fe7237a987b8b2

Note that readMSH is not templated, instead only accepting MatrixXd vertex matrix as input.

File: src/readMSH.cpp [hunk 4]
```diff
@@ -44,25 +37,26 @@ npe_arg(filename, std::string)
...
+
   if (dtypef.type() == npe::type_f32) {
-    EigenDenseF32 v;
-    EigenDenseInt t;
-    bool ret = igl::readMSH(filename, v, t);
-    if (!ret) {
-      throw std::invalid_argument("File '" + filename + "' not found.");
-    }
-    return std::make_tuple(npe::move(v), npe::move(t));
+    EigenDenseF32 v_row_maj = v.template cast <typename EigenDenseF32::Scalar>();
```
Comment (daviesthomas) [inline src/readMSH.cpp:52]: We cast to user desired dtype for parsed vertex matrix.

File: src/remove_duplicates.cpp [hunk 5]
```diff
@@ -50,18 +53,15 @@ npe_arg(epsilon, double)
...
-  Eigen::MatrixXd nv;
-  Eigen::MatrixXi nf;
-  Eigen::VectorXi i;
-  igl::remove_duplicates(v_copy, f_copy, nv, nf, i, epsilon);
-  EigenDenseFloat nv_row_major = nv;
-  EigenDenseInt nf_row_major = nf.template cast<typename EigenDenseInt::Scalar>();
-  // FIXME: vector not allowing row major, but they should be essentially the same so i feel we can leave it as col major
-  Eigen::Matrix<typename EigenDenseInt::Scalar, Eigen::Dynamic, 1, Eigen::ColMajor> i_row_major = i.template cast<typename EigenDenseInt::Scalar>();
-  return std::make_tuple(npe::move(nv_row_major), npe::move(nf_row_major), npe::move(i_row_major));
+  Eigen::MatrixXd sv;
```
Comment (daviesthomas) [inline src/remove_duplicates.cpp:59]: @teseoch I could use guidance on how to refactor this function. remove_duplicates is no longer present in libigl. 

I see three options here: 
1. deprecate this function instead redirecting users to remove_duplicate_vertices 
2. keep this function and directly compute the currently missing "I" output (count of duplicates for each face)
3. keep this function but remove support for "I" parameter... 

Let me know how to proceed here, I'm unsure how users might be using this today.

File: src/write_triangle_mesh.cpp [hunk 6]
```diff
@@ -41,15 +42,20 @@ npe_arg(v, dense_float, dense_double)
...
 npe_begin_code()
   assert_valid_3d_tri_mesh(v, f);
   // TODO: remove __copy
   //copy is necessary for the ply library
   Eigen::MatrixXi f_copy = f.template cast<int>();
   Eigen::MatrixXd v_copy = v.template cast<double>();
-  bool ok = igl::write_triangle_mesh(str, v_copy, f_copy, force_ascii);
+  
+  igl::FileEncoding encoding = igl::FileEncoding::Ascii;
+  if (!force_ascii) {
```
Comment (daviesthomas) [inline src/write_triangle_mesh.cpp:53]: "write_triangle_mesh" no longer has parameter "force_ascii" instead it takes in the enum FileEncoding. To not break old usage of our wrapper I simply assume if they are not forcing ascii, then they are forcing binary.

Comment (daviesthomas) [reply src/write_triangle_mesh.cpp:53]: A better method would be to update our API to also use the new enum... but that is a breaking change 👎 

PR that lead to this here: https://github.com/libigl/libigl/pull/1619

**Review by daviesthomas — COMMENTED** [review]
@teseoch This PR just became a whole lot more broad... If you want I can separate this into two PRs. One getting python-bindings up to date with commit BEFORE my fastwindingnumber commit on libigl (https://github.com/libigl/libigl/pull/1630) and another for my simple sign_type extension... 

For now I'll leave as is.

File: src/is_border_vertex.cpp (hunk 2 above)
Comment (teseoch) [reply src/is_border_vertex.cpp:43]: yup

File: cmake/PyiglDownloadExternal.cmake (hunk 1 above)
Comment (teseoch) [reply cmake/PyiglDownloadExternal.cmake:30]: yes thanks

File: src/readMSH.cpp (hunk 4 above)
Comment (teseoch) [reply src/readMSH.cpp:52]: perfect

File: src/remove_duplicates.cpp (hunk 5 above)
Comment (teseoch) [reply src/remove_duplicates.cpp:57]: i think opt 2 is better

File: src/write_triangle_mesh.cpp (hunk 6 above)
Comment (teseoch) [reply src/write_triangle_mesh.cpp:53]: i think in python is more clear, you can always to `force_ascii=True`
let's keep it as is

**Comment by teseoch** [comment]
thank you so much for all the changes and updates (I didn't realize that igl had so many breaking changes).

It doesn't matter to have 1/2 PR. Do you know how to fix the `remove_duplicate` to return `I`?
If not, I can do it

**Comment by daviesthomas** [comment]
> thank you so much for all the changes and updates (I didn't realize that igl had so many breaking changes).
> 
> It doesn't matter to have 1/2 PR. Do you know how to fix the `remove_duplicate` to return `I`?
> If not, I can do it

I do not know how :) I can learn or I'm (obviously) okay if you want to fix.

**Comment by teseoch** [comment]
Ok, I looked at the documentation, and it seems that the parameter I was not documented. I pushed a change that removes it.

As soon as it passes the CI, I will merge.

Thank you again!

**Pull request merged**